signature/jws/jwt.go

```diff
@@ -0,0 +1,109 @@
+package jws
+
+import (
+	"crypto/x509"
+	"fmt"
+
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/notaryproject/notation-core-go/signature"
+)
+
+// remoteSigningMethod wraps the external signer to be a jwt.SigningMethod
+type remoteSigningMethod struct {
+	jwt.SigningMethod
+	signer signature.Signer
+}
+
+func (s *remoteSigningMethod) Sign(signingString string, key interface{}) (string, error) {
+	keySpec, err := s.signer.KeySpec()
+	if err != nil {
+		return "", err
+	}
+
+	// get hasher
+	hasher := keySpec.SignatureAlgorithm().Hash()
+	if !hasher.Available() {
+		return "", &signature.SignatureAlgoNotSupportedError{
+			Alg: fmt.Sprintf("#%d", hasher)}
+	}
+
+	// calculate hash
+	h := hasher.New()
+	h.Write([]byte(signingString))
+	hash := h.Sum(nil)
+
+	// sign by external signer
+	sig, err := s.signer.Sign(hash)
+	return string(sig), err
+}
+
+// verifyJWT verifies the JWT token against the specified verification key
+func verifyJWT(tokenString string, cert *x509.Certificate) error {
+	keySpec, err := signature.ExtractKeySpec(cert)
+	if err != nil {
+		return err
+	}
+	jwsAlg, err := getJWSAlgo(keySpec.SignatureAlgorithm())
+	if err != nil {
+		return err
+	}
+	signingMethod := jwt.GetSigningMethod(jwsAlg)
+
+	// parse and verify token
+	parser := &jwt.Parser{
+		ValidMethods:         []string{"PS256", "PS384", "PS512", "ES256", "ES384", "ES512"},
+		UseJSONNumber:        true,
+		SkipClaimsValidation: true,
+	}
+
+	if _, err := parser.Parse(tokenString, func(t *jwt.Token) (interface{}, error) {
+		if t.Method.Alg() != signingMethod.Alg() {
+			return nil, &signature.MalformedSignatureError{
+				Msg: fmt.Sprintf("unexpected signing method: %v: require %v", t.Method.Alg(), signingMethod.Alg())}
+		}
+
+		// override default signing method with key-specific method
+		t.Method = signingMethod
+		return cert.PublicKey, nil
+	}); err != nil {
+		return &signature.SignatureIntegrityError{Err: err}
+	}
+	return nil
+}
+
+func getJwsAlgorithm(signer signature.Signer) (string, error) {
+	keySpec, err := signer.KeySpec()
+	if err != nil {
+		return "", err
+	}
+	return getJWSAlgo(keySpec.SignatureAlgorithm())
+}
+
+func (s *remoteSigningMethod) Alg() string {
+	alg, err := getJwsAlgorithm(s.signer)
+	if err != nil {
+		panic(err)
```

Comment:
Can we return an error instead of calling panic and crash the whole process?

Reply:
I cannot return an error, because remoteSigningMethod implements this interface in jwt package:

> // SigningMethod can be used add new methods for signing or verifying tokens.
type SigningMethod interface {
	Verify(signingString, signature string, key interface{}) error // Returns nil if signature is valid
	Sign(signingString string, key interface{}) (string, error)    // Returns encoded signature or error
	Alg() string                                                   // returns the alg identifier for this method (example: 'HS256')
}